File: src/main/java/pro/sky/java/weatherapp/service/HistoryServiceImpl.java
package pro.sky.java.weatherapp.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import pro.sky.java.weatherapp.component.HistoryRestClient;
import pro.sky.java.weatherapp.domain.HistoryRecord;
import reactor.core.publisher.Mono;

@Slf4j
@Service
@RequiredArgsConstructor
public class HistoryServiceImpl implements HistoryService {

    private final HistoryRestClient historyRestClient;

    @Override
    public Mono<HistoryRecord> save(HistoryRecord record) {
        log.info("Trying to save history record: {}", record);
        return historyRestClient.saveRecord(record)
                .doOnError(e -> log.warn("Can't save history record: {}", record, e))
                .doOnSuccess(res -> log.info("History record is saved: {}", res));
    }
}

File: src/main/kotlin/pro/sky/java/weatherapp/service/HistoryServiceImpl.kt
package pro.sky.java.weatherapp.service

import org.springframework.stereotype.Service
import pro.sky.java.weatherapp.component.HistorySaver
import pro.sky.java.weatherapp.domain.HistoryRecord

@Service
class HistoryServiceImpl(
    private val historySaver: HistorySaver
) : HistoryService {

    override suspend fun save(record: HistoryRecord) {
        return historySaver.saveRecord(record)
    }
}